### PR TITLE
🔧 Practical and Descriptive Fix artifactId to filename matching for accurate purl generation in Java archives

### PR DESCRIPTION
**Pull request by raghavvag**
Here’s a complete PR description that you can use when submitting your pull request. It includes the issue context, motivation, the fix implemented, and testing details, all properly formatted for open-source contributions.

🛠️ What this PR Does
This PR fixes an issue where Java artifacts (especially transitive dependencies like spring-ldap-core) were being incorrectly matched or misidentified in the generated SBOMs, leading to incomplete or incorrect purl values and missing vulnerabilities when scanning with Grype.

🎯 The Problem
When running syft against a Java project using:

bash
Copy
Edit
syft dir:. --exclude ./**/sbom.*.json --output cyclonedx-json@1.5=target/sbom.cyclonedx.test.json
with a dependency such as:

xml
Copy
Edit
<dependency>
  <groupId>org.springframework.ldap</groupId>
  <artifactId>spring-ldap-core</artifactId>
  <version>3.1.10</version>
</dependency>
The SBOM was generated with an incorrect purl:

json
Copy
Edit
"purl": "pkg:maven/spring-ldap-core/spring-ldap-core@3.1.4"
This caused Grype to miss known vulnerabilities such as GHSA-mqvr-2rp8-j7h4 (reported for org.springframework.ldap:spring-ldap-core).

✅ What Changed
✅ Artifact ID to Filename Matching Logic
Improved the matchArtifactIDToFilename logic to handle cases where:

Filename includes version suffixes (e.g., spring-ldap-core-3.1.4.jar)

The artifactId may appear as a prefix or suffix in the filename

Version patterns (like -1.2.3, _1.2.3, .1.2.3) are now stripped for matching

✅ New Utility
Introduced removeVersionSuffix() to normalize and remove common version patterns from filenames and artifact IDs for more reliable comparisons.

✅ Unit Tests
Added extensive test cases to cover:

Exact matches

Matches after removing version suffixes

Underscore or dot-separated version patterns

Mismatches

Validation of correct purl generation using groupId

🔍 Reproducible Issue
Before fix, Syft generated:

json
Copy
Edit
"purl": "pkg:maven/spring-ldap-core/spring-ldap-core@3.1.4"
Expected (now fixed):

json
Copy
Edit
"purl": "pkg:maven/org.springframework.ldap/spring-ldap-core@3.1.4"
This now correctly triggers CVE reporting from Grype.

🧪 How It Was Tested
Unit tests added to archive_parser_test.go using:

Artifact + filename pairings

Simulated filename inputs with various versioning patterns

Confirmed correct purl generation via:

Test_springLdapCorrectPurlGeneration

Validated SBOM output now includes correct groupId + artifactId

Grype now correctly reports vulnerabilities using fixed purl

🧾 Related Issue
Fixes: https://github.com/anchore/syft/issues/4030

📌 Type of Change
✅ Bug fix (non-breaking change which fixes an issue)

✅ Test coverage added

📋 Checklist
 Unit tests added for all core logic

 Verified output SBOM and purl

 Grype correctly picks up vulnerabilities

 Comments and cleanup complete

### CODE REVIEW

**Comment by raghavvag** [comment]
Hi @maintainers,

This PR fixes the incorrect `purl` generation issue for Java artifacts (e.g., `spring-ldap-core`) by improving artifactId-to-filename matching logic. It addresses [#4030](https://github.com/anchore/syft/issues/4030).

I've added unit tests and verified that the vulnerability detection now works as expected. Could you please:

- ✅ Review the PR for approval
- ⚙️ Approve the pending workflows (since this PR is from a fork)
- 🔁 Let me know if anything else is needed

Thanks!

File: syft/pkg/cataloger/java/archive_parser.go [hunk 1]
```diff
@@ -433,14 +434,53 @@ func artifactIDMatchesFilename(artifactID, fileName string, artifactsMap map[str
...
+
+	// Try matching with the original logic as fallback
 	return strings.HasPrefix(artifactID, fileName) || strings.HasSuffix(fileName, artifactID)
 }
 
+// removeVersionSuffix removes common version patterns from artifact names to improve matching
+func removeVersionSuffix(name string) string {
+	// Remove patterns like: -1.2.3, _1.2.3, .1.2.3
+	versionPatterns := []string{
+		`-\d+(\.\d+)*.*$`,
```
Comment (kzantow) [inline syft/pkg/cataloger/java/archive_parser.go]: This could be consolidated to a single pattern using the character set: `[-_.]`, e.g. `[-_.]\d+(\.\d+)*$` and I've removed the `.*` at the end will result in it matching things like `spring-2something`, which doesn't seem intended. I think we're going to want to be very careful making changes here, while this Java JAR determination logic may seem confusing, it's been refined a lot over the years for many specific cases.

I think we should add a test for something like the Kafka jars: https://repo1.maven.org/maven2/org/apache/kafka/kafka_2.10/0.10.2.0/ -- note these have `_2.10`, which is part of the package name, followed by a version of `-0.10.2.0`. Using the regexes defined in this PR, `_2.10-0.10.2.0` would match the version, whereas adjusting it as I've noted in this comment would not.

**Comment by raghavvag** [comment]
Closing this PR to create a new one with a more focused solution that specifically addresses the Kafka JAR parsing issue mentioned in the review feedback. The new PR will have:

- Dedicated fix for kafka_2.10-0.10.2.0 pattern handling
- Comprehensive test coverage for edge cases
- Clear commit message addressing reviewer concerns

New PR coming shortly with the improved implementation.
